src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,549 @@
+                root = _service.ChangeNamespaceDeclaration((TCompilationUnitSyntax)root, _oldNamespaceParts, _newNamespaceParts)
+                    .WithAdditionalAnnotations(Formatter.Annotation);
+
+                // Need to invoke formatter explicitly since we are doing the diff merge ourselves.
+                root = await Formatter.FormatAsync(root, Formatter.Annotation, documentWithAddedImports.Project.Solution.Workspace, optionSet, cancellationToken)
+                    .ConfigureAwait(false);
+
+                root = root.WithAdditionalAnnotations(Simplifier.Annotation);
+                var formattedDocument = documentWithAddedImports.WithSyntaxRoot(root);
+                return await Simplifier.ReduceAsync(formattedDocument, optionSet, cancellationToken).ConfigureAwait(false);
+            }
+
+            private async Task<Document> FixReferencingDocumentAsync(
+                Document document,
+                IEnumerable<ReferenceLocation> refLocations,    
+                CancellationToken cancellationToken)
+            {
+                // 1. Fully qualify all simple references (i.e. not via an alias) with new namespace.
+                // 2. Add using of new namespace (for each reference's container).
+                // 3. Try to simplify qualified names introduced from step(1).
+
+                var addImportService = document.GetLanguageService<IAddImportsService>();
+                var syncNamespaceService = document.GetLanguageService<ISyncNamespaceService>();
+                
+                var (documentWithRefFixed, containers) = 
+                    await FixReferencesAsync(document, addImportService, syncNamespaceService, refLocations, _newNamespaceParts, cancellationToken)
+                        .ConfigureAwait(false);
+
+                var optionSet = await documentWithRefFixed.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+                var placeSystemNamespaceFirst = optionSet.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, documentWithRefFixed.Project.Language);
+
+                var documentWithAdditionalImports = await AddImportsInContainersAsync(
+                    documentWithRefFixed, 
+                    addImportService,
+                    containers,
+                    ImmutableArray.Create(_state.TargetNamespace), 
+                    placeSystemNamespaceFirst, 
+                    cancellationToken).ConfigureAwait(false);
+
+                // Need to invoke formatter explicitly since we are doing the diff merge ourselves.
+                var formattedDocument = await Formatter.FormatAsync(documentWithAdditionalImports, Formatter.Annotation, optionSet, cancellationToken)
+                    .ConfigureAwait(false);
+
+                return await Simplifier.ReduceAsync(formattedDocument, optionSet, cancellationToken).ConfigureAwait(false);
```

Comment:
so you do use reduce, just doesn't use expand? is it because that's expensive and you know that it is simple to just add fully qualified namespace?